### PR TITLE
plan: let "allocID()" return int to make performance happy

### LINKED ISSUE #4440: `newBasePlan()` takes a lot of time
a lot of time is spent on `allocID()` and `runtime.concatstring2`. in function `allocID()`, much time is spent on `fmt.Sprintf()`

maybe we can let `allocID()` just return the `int` ID, and do `fmt.Sprintf()` outside, this can leave out a lot of `runtime.concatstring2` function calls

### PR DESCRIPTION
**Pull request by zz-jason**
fix #4440

### CODE REVIEW

File: plan/logical_plan_builder.go [hunk 1]
```diff
@@ -1345,7 +1345,7 @@ func (b *planBuilder) buildSemiApply(outerPlan, innerPlan LogicalPlan, condition
 	join := b.buildSemiJoin(outerPlan, innerPlan, condition, asScalar, not)
 	ap := &LogicalApply{LogicalJoin: *join}
 	ap.tp = TypeApply
-	ap.id = ap.tp + ap.allocator.allocID()
+	ap.id = fmt.Sprintf("%s_%v", ap.tp, ap.allocator.allocID())
```
Comment (ngaut) [inline plan/logical_plan_builder.go]: Can we just call alloID()? since ap.tp is alwarys TypeApply.

File: plan/plan.go [hunk 2]
```diff
@@ -371,7 +371,7 @@ func newBasePlan(tp string, allocator *idAllocator, ctx context.Context, p Plan)
 	return &basePlan{
 		tp:        tp,
 		allocator: allocator,
-		id:        tp + allocator.allocID(),
+		id:        fmt.Sprintf("%s_%v", tp, allocator.allocID()),
```
Comment (ngaut) [inline plan/plan.go]: Can you set the type of id to int?

File: plan/logical_plan_builder.go [hunk 3]
```diff
@@ -1345,7 +1345,7 @@ func (b *planBuilder) buildSemiApply(outerPlan, innerPlan LogicalPlan, condition
 	join := b.buildSemiJoin(outerPlan, innerPlan, condition, asScalar, not)
 	ap := &LogicalApply{LogicalJoin: *join}
```
Comment (winoros) [inline plan/logical_plan_builder.go]: Here can call init() directly?

File: plan/logical_plan_builder.go [hunk 4]
```diff
@@ -1343,10 +1343,9 @@ func (b *planBuilder) buildSemiApply(outerPlan, innerPlan LogicalPlan, condition
 	b.optFlag = b.optFlag | flagBuildKeyInfo
 	b.optFlag = b.optFlag | flagDecorrelate
 	join := b.buildSemiJoin(outerPlan, innerPlan, condition, asScalar, not)
-	ap := &LogicalApply{LogicalJoin: *join}
-	ap.tp = TypeApply
-	ap.id = ap.tp + ap.allocator.allocID()
-	ap.self = ap
+	ap := LogicalApply{
+		LogicalJoin: *join,
```
Comment (winoros) [inline plan/logical_plan_builder.go]: `LogicalJoin` should be assigned after call `init()`

File: expression/constant_test.go [hunk 5]
```diff
@@ -30,10 +31,10 @@ var _ = Suite(&testExpressionSuite{})
 
 type testExpressionSuite struct{}
 
-func newColumn(name string) *Column {
+func newColumn(from int) *Column {
 	return &Column{
-		FromID:  name,
-		ColName: model.NewCIStr(name),
+		FromID:  from,
+		ColName: model.NewCIStr(fmt.Sprintf("%v", from)),
```
Comment (hanfei1991) [inline expression/constant_test.go]: Sprint(from) is enough

File: plan/plan.go [hunk 6]
```diff
@@ -54,7 +54,11 @@ type Plan interface {
 	// Get the schema.
 	Schema() *expression.Schema
 	// Get the ID.
-	ID() string
+	ID() int
+	// Get the ID in explain statement
+	ExplainID() string
+	// Get the type.
+	Type() string
```
Comment (hanfei1991) [inline plan/plan.go]: What is `Type()` used for ?

File: plan/logical_plan_builder.go (hunk 4 above)
Comment (zz-jason) [reply plan/logical_plan_builder.go]: we'd better not call `init` here. after `init`, we should `ap.LogicalJoin = *join`, then we have to realloc `id` and reset `tp`

File: plan/plan.go (hunk 6 above)
Comment (zz-jason) [reply plan/plan.go]: removed, useless code.

**Comment by hanfei1991** [comment]
LGTM

**Comment by zz-jason** [comment]
@winoros @ngaut PTAL

**Pull request merged**